app/src/main/java/dev/davidv/translator/AidlTranslationService.kt

```diff
@@ -0,0 +1,100 @@
+package dev.davidv.translator
+
+import android.app.Service
+import android.content.Intent
+import android.os.IBinder
+import android.util.Log
+import dev.davidv.translator.ITranslationCallback
+import dev.davidv.translator.ITranslationService
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+class AidlTranslationService : Service() {
+
+  private val TAG = this.javaClass.name.substringAfterLast('.')
+
+  private lateinit var translationCoordinator: TranslationCoordinator
+
+  override fun onCreate() {
+    super.onCreate()
+    val settingsManager = SettingsManager(this)
+    val filePathManager = FilePathManager(this, settingsManager.settings)
+    val translationService = TranslationService(settingsManager, filePathManager)
+    val languageDetector = LanguageDetector()
+    val imageProcessor = ImageProcessor(this, OCRService(filePathManager))
+    translationCoordinator = TranslationCoordinator(this, translationService, languageDetector, imageProcessor, settingsManager, false)
+    Log.d(TAG, "onCreate")
+  }
+
+  override fun onBind(intent: Intent?): IBinder {
+    Log.d(TAG, "onBind")
+    return binder
+  }
+
+  private val binder = object : ITranslationService.Stub() {
+    override fun translate(textToTranslate: String?, fromLanguageStr: String?, toLanguageStr: String?, callback: ITranslationCallback?) {
+      Log.d(TAG, "txt len:${textToTranslate?.length ?: -1}, from:$fromLanguageStr, to:$toLanguageStr, cb = ${callback != null}")
+
+      if (textToTranslate == null || callback == null) {
+        Log.w(TAG, "translate: textToTranslate or callback is null")
+        return
+      }
+
+      val fromLanguage = fromLanguageStr?.takeIf { it.isNotEmpty() }?.let { lng -> Language.entries.find { it.code == lng } }
+      val toLanguage = toLanguageStr?.takeIf { it.isNotEmpty() }?.let { lng -> Language.entries.find { it.code == lng } }
+
+      synchronized (this) {
```

Comment:
Why does this hold a lock but launch its work into a coroutine? I don't think the lock is held during the translation

Reply:
I was trying to fix the problem where it seem to get stuck when there's too many translate requests. I'll test some more with the lock removed later.

Reply:
Yeah I definitely noticed this on the app itself, just need to rework the lock, I think

Reply:
Actually I have a while loop checking if it's `isTranslating` just a bit below than the marked lines, hopefully it's enough.